edgelet/edgelet-config/src/yaml_file_source.rs

```diff
@@ -0,0 +1,133 @@
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::PathBuf;
+
+use config::{ConfigError, Source, Value};
+use yaml_rust::{Yaml, YamlLoader};
+
+/// This is similar to [`config::File`] in that it parses a YAML string or file and implements `config::Source`.
+///
+/// We use this to parse our config.yaml instead of `config::File` because `config::File` lower-cases all field names that it reads.
+/// This causes issues with fields like `agent.config.createOptions` since the config crate returns `agent.config.createoptions`
+/// which the serde deserializer ignores.
+#[derive(Clone, Debug)]
+pub(crate) enum YamlFileSource {
+    File(PathBuf),
+    String(&'static str),
+}
+
+impl Source for YamlFileSource {
+    fn clone_into_box(&self) -> Box<dyn Source + Send + Sync> {
+        Box::new(self.clone())
+    }
+
+    fn collect(&self) -> Result<HashMap<String, Value>, ConfigError> {
+        let origin = match self {
+            YamlFileSource::File(path) => Some(path.to_string_lossy().into_owned()),
+            YamlFileSource::String(_) => None,
+        };
+
+        let s = match self {
+            YamlFileSource::File(path) => {
+                let mut file =
+                    File::open(path).map_err(|err| ConfigError::Foreign(Box::new(err)))?;
+                let mut contents = String::new();
+                let _ = file
+                    .read_to_string(&mut contents)
+                    .map_err(|err| ConfigError::Foreign(Box::new(err)))?;
+                Cow::Owned(contents)
+            }
+
+            YamlFileSource::String(s) => Cow::Borrowed(*s),
+        };
+
+        let docs =
+            YamlLoader::load_from_str(&*s).map_err(|err| ConfigError::Foreign(Box::new(err)))?;
+
+        let mut docs = docs.into_iter();
+        let doc = match docs.next() {
+            Some(doc) => {
+                if docs.next().is_some() {
+                    return Err(ConfigError::Foreign(Box::new(std::io::Error::new(
+                        std::io::ErrorKind::Other,
```

Comment:
Would InvalidInput be better if the user provided these yaml docs?

Reply:
I should probably just not be lazy and make an actual error type.